### PR TITLE
push down cleanup in order not to delete pod when there is no execution id

### PR DESCRIPTION
**Pull request by honnix**
@danielnorberg a quick fix by pushing down cleanup action

### CODE REVIEW

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 1]
```diff
@@ -402,11 +399,17 @@ private synchronized void tryPollPods() {
...
+  }
+
+  private Optional<RunState> lookupPodRunState(Pod pod, WorkflowInstance workflowInstance,
+                                               Runnable cleanupAction) {
     final String podName = pod.getMetadata().getName();
 
     final RunState runState = stateManager.get(workflowInstance);
     if (runState == null) {
       LOG.warn("Pod event for unknown or inactive workflow instance {}", workflowInstance);
+      cleanupAction.run();
```
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: call cleanup directly instead?

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 2]
```diff
@@ -420,6 +423,7 @@ private synchronized void tryPollPods() {
     if (!podName.equals(executionId)) {
       LOG.warn("Pod event not matching current exec id, current:{} != pod:{}",
           executionId, podName);
+      cleanupAction.run();
```
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: same as above

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 3]
```diff
@@ -378,12 +378,9 @@ private synchronized void tryPollPods() {
       if (!workflowInstance.isPresent()) {
         continue;
       }
-      final Optional<RunState> runState = lookupPodRunState(pod, workflowInstance.get());
-      if (runState.isPresent()) {
-        emitPodEvents(Watcher.Action.MODIFIED, pod, runState.get());
-      } else {
-        cleanup(workflowInstance.get(), pod.getMetadata().getName());
-      }
+      lookupPodRunState(pod, workflowInstance.get(), () -> cleanup(workflowInstance.get(),
```
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: this code looks a bit confusing now

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 1 above)
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: this method is shared by both polling and watcher.

**Comment by danielnorberg** [comment]
An alternative solution would be to just have a cleanup grace period. I.e., only remove pod if creationTImestamp more than X minutes ago.

**Comment by danielnorberg** [comment]
👍 but I guess we'll want to revisit this soon =)

**Comment by honnix** [comment]
yeah, we should probably avoid this race condition at all.

**Comment by honnix** [comment]
https://github.com/spotify/styx/blob/a90f9b5b8fa76a1be71ffceb7c72e899a33559ab/styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java#L76-L76
😆

**Comment by danielnorberg** [comment]
palm, meet face

**Pull request merged**